hypershift-operator/controllers/nodepool/conditions.go

```diff
@@ -824,3 +826,107 @@ func (r NodePoolReconciler) createValidGeneratedPayloadCondition(ctx context.Con
 
 	return condition, nil
 }
+
+func (r *NodePoolReconciler) supportedVersionSkewCondition(ctx context.Context, nodePool *hyperv1.NodePool, hcluster *hyperv1.HostedCluster) (*ctrl.Result, error) {
+	if hcluster.Status.Version == nil {
+		SetStatusCondition(&nodePool.Status.Conditions, hyperv1.NodePoolCondition{
+			Type:               hyperv1.NodePoolSupportedVersionSkewConditionType,
+			Status:             corev1.ConditionFalse,
+			Reason:             hyperv1.NodePoolValidationFailedReason,
+			Message:            "HostedCluster version is not available yet, waiting for version information",
+			ObservedGeneration: nodePool.Generation,
+		})
+		return &ctrl.Result{}, fmt.Errorf("HostedCluster version is not available yet")
+	}
+	removeStatusCondition(&nodePool.Status.Conditions, hyperv1.NodePoolSupportedVersionSkewConditionType)
+
+	// Get the control plane version string
+	var controlPlaneVersionStr string
+	if len(hcluster.Status.Version.History) == 0 {
+		// If the cluster is in the process of installation, there is no history
+		// Use the desired version as the control plane version
+		controlPlaneVersionStr = hcluster.Status.Version.Desired.Version
+	} else {
+		// If the cluster is installed or upgrading
+		// Start with the most recent version from history as the default
+		controlPlaneVersionStr = hcluster.Status.Version.History[len(hcluster.Status.Version.History)-1].Version
+		// Update with any more recent Completed version if found
+		for _, history := range hcluster.Status.Version.History {
+			if history.State == "Completed" {
+				controlPlaneVersionStr = history.Version
+				break
+			}
+		}
+	}
+	// Parse control plane version
+	controlPlaneVersion, err := semver.Parse(controlPlaneVersionStr)
+	if err != nil {
+		return &ctrl.Result{}, fmt.Errorf("parsing control plane version (%s): %w", controlPlaneVersionStr, err)
+	}
+
+	releaseImage, err := r.getReleaseImage(ctx, hcluster, nodePool.Status.Version, nodePool.Spec.Release.Image)
+	if err != nil {
+		SetStatusCondition(&nodePool.Status.Conditions, hyperv1.NodePoolCondition{
```

Comment:
why are we setting condition in some errors but not in others?

Reply:
From my perspective, the error from semver.Parse() seems more like a rare or accidental issue that would be unlikely to occur during actual execution. So I didn’t add a Condition for it. I'm not sure if my understanding is correct. I noticed that in some cases in this section of the code: https://github.com/openshift/hypershift/blob/1c615c778e4e321819827a78accd9ecd2199abcc/hypershift-operator/controllers/nodepool/conditions.go#L315-L326, Conditions aren’t set either.

Reply:
Agree that is unlikely to happen but also does not harm to be added as a msg of the condition in case of that parsing fails, it's a more deterministic approach.